client/tw/categorical.ts

```diff
@@ -0,0 +1,271 @@
+
+			2. The fill() function must fill-in any expected missing values,
+			   validate the tw.q shape at runtime, and throw on any error or mismatched expectation.
+			   Runtime validation is required because the input raw tw can come from anywhere,
+			   like term.bins.default, which is a runtime variable that is not possible to statically check.
+
+			3. The filled-in tw, when returned, must be **coerced** to the full CatTW* type, 
+			   in order to match the function signature's return type.
+		*/
+
+		/*
+			For each of fill() functions below:
+			1. The `tw` argument must already have a tw.type string value, 
+			   which corresponds to the RawNumTW* equivalent of the full NumTW* type.
+			   NOTE: This only applies for routing, it's optional when calling the 
+			   specialized fill() function directly.
+
+			2. The fill() function must fill-in any expected missing values,
+			   validate the tw.q shape at runtime, and throw on any error or mismatched expectation.
+			   Runtime validation is required because the input raw tw can come from anywhere,
+			   like term.bins.default, which is a runtime variable that is not possible to statically check.
+
+			3. The filled-in tw, when returned, must be **coerced** to the full NumTW* type, 
+			   in order to match the function signature's return type.
+		*/
+		switch (tw.type) {
+			case 'CatTWValues':
+				return CatValues.fill(tw)
+
+			case 'CatTWPredefinedGS':
+				return CatPredefinedGS.fill(tw, opts)
+
+			case 'CatTWCustomGS':
+				return CatCustomGS.fill(tw)
+
+			default:
+				throw `tw.type='${tw.type} (q.mode:q.type=${tw.q.mode}:${tw.q.type}' is not supported by CategoricalBase.fill()`
+		}
+	}
+}
+
+export class CatValues extends TwBase {
+	term: CategoricalTerm
+	q: ValuesQ
+	#tw: CatTWValues
+	#opts: TwOpts
+
+	// declare a constructor, to narrow the tw type
+	constructor(tw: CatTWValues, opts: TwOpts = {}) {
+		super(tw, opts)
+		this.term = tw.term
+		this.q = tw.q
+		this.#tw = tw
+		this.#opts = opts
+	}
+
+	// See the relevant comments in the CategoricalBase.fill() function above
+	static fill(tw: RawCatTWValues, opts: TwOpts = {}): CatTWValues {
+		if (!tw.type) tw.type = 'CatTWValues'
+		else if (tw.type != 'CatTWValues') throw `expecting tw.type='CatTWValues', got '${tw.type}'`
+
+		if (tw.term.type != 'categorical') throw `expecting tw.term.type='categorical', got '${tw.term.type}'`
+		const { term, q } = tw
+		if (!q.type) q.type = 'values'
+		else if (q.type != 'values') throw `expecting tw.q.type='values', got ${tw.q.type}`
+
+		// GDC or other dataset may allow missing term.values
+		if (!term.values) term.values = {}
+		const numVals = Object.keys(tw.term.values).length
+		// GDC or other dataset may allow empty term.values
+		//if (!numVals) throw `empty term.values`
+		if (q.mode == 'binary') {
+			if (Object.keys(tw.term.values).length != 2) throw 'term.values must have exactly two keys'
+			// TODO:
```

Comment:
`tw.term.values` should not be checked for `q.mode=binary`. `tw.q.customset` should be checked instead. Can address in a subsequent PR.